### DISCUSSION TITLE
path2.isCCW

### DISCUSSION THREAD

**Discussion by hrgdavor** [jscad/OpenJSCAD.org#929] [Ideas]
There is occasional need to check handiness of a path, and there is a simple function to do so, but it has it's limitations.

Since the topic needs further investigation, the function is not part of jscad but is left here for those that need it, and to allow more investigation into the problem.

If you create a path than it is better to make sure  that it is CCW before extruding, but for cases where you get a path from external source you may need to check it.
 - when reading SVG
 - when reading a font

```js
/**
 * Determin if the given path is CCW
 * @param {path2} path - the geometry to check
 * @returns {Boolean} true if the path is CCW
 * @alias module:modeling/geometries/path2.isCCW
 */
const isCCW = (geometry) => {
  let area = 0; let j
  const points = geometry.points
  const length = points.length
  points.forEach((v, i) => {
    j = (i + 1) % length
    area += v[0] * points[j][1]
    area -= points[j][0] * v[1]
  })

  return area >= 0
}

module.exports = isCCW
```

**Comment by z3dev** [comment]
An alternative would be to use area() from maths utils, which also does the same calculation. CCW handedness would be any positive area (given that the path does not cross itself.)

```js
const { path2 } = require('@jscad/modeling').geometries
const { area } = require('@jscad/modeling').maths.utils

const areaunderpath = area(path2.toPoints(mypath))
```

**Reply by hrgdavor** [reply]
Oh, did not notice there already is area calculation function. 

This could be improved, but only by knowing internals. It would be better to create a issCCW function that uses area with an additional trick. There is no need to apply transform, area can be calculated on the original geometry points without transforms, and then checked to see if transform is mirror (mat4.isMirroring).

**Reply by z3dev** [reply]
Given what you just said, isCCW() is not complete. ;)
What if the transforms are a rotation, which would position the path differently? Wouldn't that make isCCW() also incorrect?

**Reply by hrgdavor** [reply]
rotation arround Z would not change CCW, the question is if rotation arround x or y is ok to be allowed for 2d shapes.
